File: src/python/run_1d.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
from pathlib import Path

from heat.config import Config1D
from heat.solver_1d import solve


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Solve the 1-D heat equation.")
    parser.add_argument("--nx", type=int, default=Config1D.nx)
    parser.add_argument("--nt", type=int, default=Config1D.nt)
    parser.add_argument("--dt", type=float, default=Config1D.dt)
    parser.add_argument("--alpha", type=float, default=Config1D.alpha)
    parser.add_argument("--lx", type=float, default=Config1D.lx)
    parser.add_argument("--output-every", type=int, default=Config1D.output_every)
    parser.add_argument("--output-dir", type=Path, default=Config1D.output_dir)
    parser.add_argument("--initial", choices=["semicircle", "gaussian", "sine"], default=Config1D.initial)
    parser.add_argument("--boundary", choices=["dirichlet", "neumann", "periodic"], default=Config1D.boundary)
    parser.add_argument("--no-vtk", action="store_true")
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    solve(
        Config1D(
            nx=args.nx,
            nt=args.nt,
            dt=args.dt,
            alpha=args.alpha,
            lx=args.lx,
            output_every=args.output_every,
            output_dir=args.output_dir,
            initial=args.initial,
            boundary=args.boundary,
            vtk=not args.no_vtk,
        )
    )
    print(f"Calculation finished. Output directory: {args.output_dir}")


if __name__ == "__main__":
    main()

File: src/julia/run_1d.jl
include(joinpath(@__DIR__, "heat", "Heat.jl"))
using .Heat

const NX           = 32
const NT           = 2500
const OUTPUT_EVERY = 50
const DT           = 1.0e-3
const ALPHA        = 1.0
const LX           = 2.0

output_dir = get(ARGS, 1, "output/julia/1d")
initial    = get(ARGS, 2, "semicircle")
boundary   = get(ARGS, 3, "dirichlet")
method     = get(ARGS, 4, "euler")
vtk        = !(length(ARGS) >= 5 && ARGS[5] == "no-vtk")

solve_heat_1d(
    nx           = NX,
    nt           = NT,
    output_every = OUTPUT_EVERY,
    dt           = DT,
    alpha        = ALPHA,
    lx           = LX,
    output_dir   = output_dir,
    initial      = initial,
    boundary     = boundary,
    vtk          = vtk,
    method       = method,
)

println("Calculation finished. Output directory: $output_dir")
